internal/backends/winit/event_loop.rs

```diff
@@ -310,58 +294,34 @@ fn process_window_event(
                 _ => None,
             });
             if let Some(text) = input.virtual_keycode.and_then(key_codes::winit_key_to_string) {
-                #[allow(unused_mut)]
-                let mut modifiers = window.current_keyboard_modifiers().get();
-                // On wasm, the WindowEvent::ModifiersChanged event is not received
-                #[cfg(target_arch = "wasm32")]
-                #[allow(deprecated)]
-                {
-                    modifiers.shift |= input.modifiers.shift();
-                    modifiers.control |= input.modifiers.ctrl();
-                    modifiers.meta |= input.modifiers.logo();
-                    modifiers.alt |= input.modifiers.alt();
-                }
-
-                let event = key_event(
-                    match input.state {
+                runtime_window.process_key_input(KeyInputEvent {
+                    event_type: match input.state {
                         winit::event::ElementState::Pressed => KeyEventType::KeyPressed,
                         winit::event::ElementState::Released => KeyEventType::KeyReleased,
                     },
                     text,
-                    modifiers,
-                );
-                runtime_window.process_key_input(&event);
+                    ..Default::default()
+                });
             };
         }
         WindowEvent::Ime(winit::event::Ime::Preedit(string, preedit_selection)) => {
             let preedit_selection = preedit_selection.unwrap_or((0, 0));
-            let event = KeyEvent {
+            let event = KeyInputEvent {
                 event_type: KeyEventType::UpdateComposition,
                 text: string.into(),
                 preedit_selection_start: preedit_selection.0,
                 preedit_selection_end: preedit_selection.1,
                 ..Default::default()
             };
-            runtime_window.process_key_input(&event);
+            runtime_window.process_key_input(event);
         }
         WindowEvent::Ime(winit::event::Ime::Commit(string)) => {
-            let event = KeyEvent {
+            let event = KeyInputEvent {
                 event_type: KeyEventType::CommitComposition,
                 text: string.into(),
                 ..Default::default()
             };
-            runtime_window.process_key_input(&event);
-        }
-        WindowEvent::ModifiersChanged(state) => {
```

Comment:
Shouldn't we still listen to this winit event and translate it to KeyPress/Release events?

Reply:
@tronical: Winit already gives key event for these keys.  no need to duplicate

Reply:
I've added `ControlL` and `ShiftL` to the `key_codes` with the same unicode. It's fulfills the specification of https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode. It worked for the modifieres with winti and qt backend.

Reply:
It's good that `ControlL` and `ShiftL` are now in `key_codes`. I think the next step is to detect their use in `state_update` and map them to the `control` and `shift` modifiers (modulo the macOS swap).